torch/_inductor/lowering.py

```diff
@@ -4742,10 +4742,215 @@ def max_pool3d_with_indices(
+                kernel_offset = indices_loader(idx)
+
+                # Convert to 2D kernel position with bounds checking
+                kernel_h_idx = kernel_offset // kernel_size[1]
+                kernel_w_idx = kernel_offset % kernel_size[1]
+
+                # Validate kernel indices are within bounds
+                kernel_h_valid = ops.and_(
+                    ops.ge(kernel_h_idx, 0), ops.lt(kernel_h_idx, kernel_size[0])
+                )
+                kernel_w_valid = ops.and_(
+                    ops.ge(kernel_w_idx, 0), ops.lt(kernel_w_idx, kernel_size[1])
+                )
+                kernel_valid = ops.and_(kernel_h_valid, kernel_w_valid)
+
+                # Calculate absolute input tensor positions from output coordinates
+                h_base = out_h_idx * stride[0] - padding[0]
+                h_offset = kernel_h_idx * dilation[0]
+                abs_h = h_base + h_offset
+
+                w_base = out_w_idx * stride[1] - padding[1]
+                w_offset = kernel_w_idx * dilation[1]
+                abs_w = w_base + w_offset
+
+                # Comprehensive bounds validation
+                h_valid = ops.and_(ops.ge(abs_h, 0), ops.lt(abs_h, in_h))
+                w_valid = ops.and_(ops.ge(abs_w, 0), ops.lt(abs_w, in_w))
+                position_valid = ops.and_(h_valid, w_valid)
+
+                # Combined validation
+                all_valid = ops.and_(kernel_valid, position_valid)
+
+                # Convert 4D tensor coordinates to linear index for scatter operation
+                batch_offset = batch_idx * channels * in_h * in_w
+                channel_offset = channel_idx * in_h * in_w
+                h_offset_final = abs_h * in_w
+                linear_idx = batch_offset + channel_offset + h_offset_final + abs_w
+
+                return ops.where(all_valid, linear_idx, ops.constant(-1, torch.int64))
+            except (TypeError, AttributeError):
+                # Return sentinel value for positions that cannot be computed
+                return ops.constant(-1, torch.int64)
+
+        return ir.Pointwise.create(
+            device=indices.get_device(),
+            dtype=torch.int64,
+            inner_fn=index_transform,
+            ranges=grad_output.get_size(),
+        )
+
+    # Execute scatter-based gradient accumulation
+    try:
+        # Compute scatter indices
+        scatter_indices = compute_scatter_indices()
+
+        # Flatten tensors for scatter operation
+        grad_output_flat = view(grad_output, [grad_output.get_numel()])
+        scatter_indices_flat = view(scatter_indices, [grad_output.get_numel()])
+
+        # Initialize gradient accumulator tensor
+        grad_input_flat = _full(
+            0, grad_output.get_device(), grad_output.get_dtype(), [x.get_numel()]
+        )
+
+        # Handle invalid indices (-1) by replacing with valid index and zeroing values
+        valid_mask = ops.ne(scatter_indices_flat, ops.constant(-1, torch.int64))
```

Comment:
`ops.*` is supposed to be for used in functions passed to `inner_fn` in `ir.Pointwise.create`. I'm not sure how it works here.